File: python/a20.py
# A20 LCS
# submission: https://atcoder.jp/contests/tessoku-book/submissions/38504375

s = input()
t = input()

sl = len(s)
tl = len(t)
dp = [[0] * (tl + 1) for i in range(sl + 1)]

for i in range(sl + 1):
    for j in range(tl + 1):
        if i < sl and j < tl and s[i] == t[j]:
            dp[i + 1][j + 1] = max(dp[i + 1][j + 1], dp[i][j] + 1)
        if i < sl:
            dp[i + 1][j] = max(dp[i + 1][j], dp[i][j])
        if j < tl:
            dp[i][j + 1] = max(dp[i][j + 1], dp[i][j])

print(dp[sl][tl])

File: rust/tessoku-book/src/bin/a20.rs
// A20 LCS
// submission: https://atcoder.jp/contests/tessoku-book/submissions/38504760

use proconio::input;
use proconio::marker::Chars;

#[macro_export]
macro_rules! chmax {
    ( $a:expr, $b:expr ) => {
        {
            if $a < $b {
                $a = $b;
                true
            } else {
                false
            }
        }
    };
}

fn main() {
    input!{
        s: Chars,
        t: Chars,
    }
    let sl = s.len();
    let tl = t.len();
    let mut dp = vec![vec![0; tl + 1]; sl + 1];
    for i in 0..=sl {
        for j in 0..=tl {
            if i < sl && j < tl && s[i] == t[j] {
                chmax!(dp[i + 1][j + 1], dp[i][j] + 1);
            }
            if i < sl {
                chmax!(dp[i + 1][j], dp[i][j]);
            }
            if j < tl {
                chmax!(dp[i][j + 1], dp[i][j]);
            }
        }
    }
    println!("{}", dp[sl][tl]);
}
